### ISSUE TITLE
OpenSSL-sys build failure / Cargo MAKEFLAGS not configurable

### ISSUE THREAD

**Issue by nadenf** [rust-lang/cargo#13476]
### Problem

So we have an issue building OpenSSL-sys on Mac M1: https://github.com/sfackler/rust-openssl/issues/2179

Which is likely caused by this command:
```
running cd "/opt/actions-runner/_work/search/search/search/src-tauri/target/aarch64-apple-darwin/release/build/openssl-sys-8aca377df6729114/out/openssl-build/build/src" 
&& MAKEFLAGS="-j --jobserver-fds=7,13 --jobserver-auth=7,13" "make" "build_libs"
```

Namely that the "-j" flag for make is overloading the resources of the host.

And there doesn't seem to be a way to override this MAKEFLAGS to workaround it.

### Version

```text
cargo 1.75.0 (1d8b05cdd 2023-11-20)
release: 1.75.0
commit-hash: 1d8b05cdd1287c64467306cf3ca2c8ac60c11eb0
commit-date: 2023-11-20
host: aarch64-apple-darwin
libgit2: 1.7.1 (sys:0.18.1 vendored)
libcurl: 8.4.0 (sys:0.4.68+curl-8.4.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 1.1.1u  30 May 2023
os: Mac OS 14.2.1 [64-bit]
```

**Comment by weihanglo** [comment]
Could you provide steps to reproduce and also include the version or commit of that crate?

**Comment by nadenf** [comment]
@weihanglo  .. You can see the code and full build failure [here](https://github.com/harana/search/actions/runs/7996860525/job/21840182426) under the Build Tauri step.

Looks to be an issue both with openssl-sys-v0.9.99 and openssl-sys-v0.9.100 versions.

**Comment by weihanglo** [comment]
Is the issue about building openssl exhausting system resource?

For `MAKEFLAGS="-j --jobserver-auth=...`, it's `jobserver` crate that [passes `MAKEFLAGS="-j ..."`](https://github.com/rust-lang/jobserver-rs/blob/799eacc0e49db5f281d349da5cebd5beeb696cba/src/lib.rs#L389). By [definition](https://www.gnu.org/software/make/manual/make.html#Communicating-Options-to-a-Sub_002dmake), `-j` without a value means "to run as many jobs as possible in parallel". Given that cargo controls the max parallelism, it won't spawn more than N jobs, where N is either # of core or `-j <value>`.

However, `cc` crate has a `parallel` feature that might go beyond the limit. I've identified the potential symptom from the issue you've linked:

* `cargo-risczero` [enables the `parallel` feature](https://github.com/risc0/risc0/blob/92ef689afcce66e53d0de14cfa9b83831086c698/risc0/build_kernel/Cargo.toml#L11).
* https://github.com/tremor-rs/tremor-runtime depends on `zstd-sys`, which [enables `parallel` feature](https://github.com/gyscos/zstd-rs/blob/ed106b224b2d386a95dd54c7feb7b558ea2745d0/zstd-safe/zstd-sys/Cargo.toml#L61-L63).

If the issue is inherently about using too many system resources, then either those dependencies shouldn't enable the `parallel` feature, or `cc` crate needs to provide a way to opt-out the `parallel` feature, for example via `<pkg-name>_NO_CC_PARALLEL=true`.

(Note that features are additive and there is no cargo-native way to opt-out at this moment)

**Comment by nadenf** [comment]
@weihanglo  .. 

```
MacBook-Pro src-tauri % cargo tree -i cc -e features | grep parallel
│   └── cc feature "parallel"
└── cc feature "parallel" (*)
```

**Comment by weihanglo** [comment]
Great! Thanks for the reply.

Now, to identify if the problem is really a resource exhaustion, you can [patch](https://doc.rust-lang.org/nightly/cargo/reference/overriding-dependencies.html#the-patch-section) the `cc` crate, making `parallel` feature a no-op, and see if the error occurs in your CI.

**Comment by nadenf** [comment]
@weihanglo  .. So that fix works and have created a [PR](https://github.com/gyscos/zstd-rs/pull/262) for zstd to disable CC parallelism.

Thanks for all your help !

**Comment by nadenf** [comment]
So the fix for this is:

```
[patch.crates-io]
zstd-sys = { git = "https://github.com/harana-oss/zstd-rs.git" }
```

Until this [PR](https://github.com/gyscos/zstd-rs/pull/262) gets merged.

**Comment by NobodyXu** [comment]
Which version of cc are you using @nadenf ?

I'd like to identify the cause instead of error.

Latest 1.0.86 (compared to 1.0.79) removes thread spawning in cc when parallel is enabled, it should prevent resource exhaustion and improve performance, if that's the cause of the error.

**Comment by nadenf** [comment]
@NobodyXu  .. ZStd is using 1.0.45. I will try upgrading to 1.0.86 and leaving parallel enabled.

**Comment by nadenf** [comment]
I have tried with 1.086 and parallel enabled and it doesn't fix the problem. Only removing parallel entirely.

**Comment by NobodyXu** [comment]
Hmmm I think I know the reason.

Since cc >=1.0.80, we set the jobserver inherited to `O_NONBLOCK`, make probably isn't designed for this, which is why it failed.

I will open a new ticket in cc and fix it.

**Comment by NobodyXu** [comment]
@nadenf Can you try make >= 4.4.1 please?

I think make might have introduced support for non-vlocking jobserver pipe in newer version.

**Comment by nadenf** [comment]
@NobodyXu  .. Would add that the default on a Mac is 3.81 so you're asking developers to override that with the Homebrew version.

That said it doesn't work for me.

**Comment by NobodyXu** [comment]
Thanks, my proposed solution is to change rust-lang/jobserver to use a named pipe instead of an annoymous one.

The named pipe is passed down by its path in filesystem and opening it using `O_NONBLOCK` doesn't affect others.

We can also retain backwards compatibility by keeping the jobserver fds inheritable and keeping them in the environment variable.

**Comment by NobodyXu** [comment]
@nadenf I tried it in my own project that updates to make 4.4.1 does fix the error.
